### DISCUSSION TITLE
Attribute Values defining "Invalid values"

### DISCUSSION THREAD

**Discussion by mdeweerd** [zigpy/zigpy#760] [General]
In some cases, especially at startup, but also in case of failure, and in accordance with the specification, a Zigbee End Device (ZED) can indicate that an attribute has an "Invalid value".

For instance, the Thermostat/LocalTemperature indicates a temperature for values from 0x954d (-273.15°C) to 0x7FFF (327.67 ºC).  **A value of 0x8000 indicates that the temperature measurement is invalid,** and all other values are unused (0x8001 to 0x954C).

The same is true for several other attributes that also have a define "Invalid value".

The thing is that currently 0x8000 is interpreted as a negative value, and that offsets the graph in Home Assistant for insdtance, but could also lead to erroneous automations, etc.
Here is such a graph:
![image](https://user-images.githubusercontent.com/1504752/122709717-1ae6ee00-d25f-11eb-8fd8-29fd26238b8f.png)

The temperature starts of at 0x8000 and is valid quiet some time later (which is because of manipulations on the ZED).

As far as I can see, there is no method to declare a value as invalid in `zigpy`, and a program like `Home Assistant` should not have to know the details of Zigbee (which particular values are invalid values).

I put this up for discussion.

It looks like the best approach would be that `zigpy` simply does not record/notify `Invalid values`.

There should be some general approach for this, and it should then be (progressively) implemented for the impacted fields.  That could mean that when declaring an attribute, there should be one or more fields to indicate the valid range or the invalid value in ordert to be able to handle attributes in a generic way.

**Comment by puddly** [comment]
Interesting. I've seen references in the spec to these values but I've never encountered one myself. What sensor is sending it?

Are these invalid values specific to each cluster attribute or are they a part of the data type?

If it's the former, perhaps this can be added as a part of the ZCL v7 rewrite PR #716? I've updated the attribute definition object to hold more information (https://github.com/puddly/zigpy/blob/puddly/zclv7/zigpy/zcl/foundation.py#L624) so it shouldn't be hard to add a new field like `invalid_value`.

If it's the latter, we can add an `is_invalid` property or something to each of the core data types.

**Comment by mdeweerd** [comment]
I can't tell which sensor it is (they are) at this moment (confidential).
It happens essentially at startup or in case of a hardware error (NTC resitor measurement out of range essentially due to a shorted or missing NTC).
It also happens in a Metering cluster when the actual meter is not responding or before it responded (after startup).

Most devices I've seen should report "Invalid value" where they actually respond (with 0 for instance when they do not know the value yet).

With regards to specificity: both exist.  In the given example, this is specific to the attribute.  The Temperature Measurement uses the same rule, but it is specified specifically for the cluster attribute.  In the Power Configuration Cluster a battery voltage or percentage of 0xFF has a similar meaning (but on a byte value).  
UTCTime, a Zigbee type, specifies that 0xFFFFFFFF represents an invalid value, and this is for all attributes depending on it - so it is not repeated at the attribute level.
"Table 2-10. Datatypes" lists a few Invalid values for common data types: 0xFF is invalid for logical discretes, as well as for uint8 values (surprisingly).  UTCTime is also listed there.
Temperatures are not defined as a Data Type Class, so the invalid definitions appear on the attribute level but they are (mostly) common.  

However, considering that LocalTemperature, OutdoorTemperature, etc are specified as "int16" we can also simply observe that the Invalid value for an int16 (0x8000) simply extends to the Temperature value itself (0x8000).

So in the end it is likely that the Invalid values for the designated Data Type Class extends to the attribute even if it has a more specific unit.  

Some examination is needed for a few other fields.  "ColorTemperatureMireds" has 0xFFFF as invalid value, and 0x0000 as undefined value.  Is the "invalid value" important there to indicate that the configuration of the field is inappropriate?

The MaxPressure specification says that the value must be set to 0xFFFF (which is the Invalid value) if the maximum is not available.  In that case the Invalid value should be recorded. Several other attributes have similar constraints.

Maby a third field could be added to DATA_TYPES indicating the "Invalid value" for the type.  There could be a getter that only returns valid values, and a getter that returns any value., and a is_valid (and/or is_invalid) property.

**Reply by MattWestb** [reply]
I think this is very interesting but shall Zigppy and its radio libs caring of it or only passing the information to ZHA that shall interpreting the values and saying its one valid or not for presenting and logging ?

If patching in Zigpy that is also many cases with light like 0x0 and 0xffff is normally not one valid light level.

**Reply by puddly** [reply]
All of the types in `DATA_TYPES` correspond one-to-one with basic data types. Maybe it would be best to directly extend all of the basic types with the info contained in that dict?

```Python
class uint8_t(uint_t, bits=8):
    type_id = 0x20
    type_name = "Unsigned Integer"
    type_sampling = Discrete

    invalid_values = (0xFF,)
```

**Reply by Adminiuga** [reply]
That kind of logic is best suited for the application level, since the application does some sort of value "cooking" anyway.

**Reply by Adminiuga** [reply]
And yeah, zha should have done this long time ago, just until now haven't seen any device behaving as per specs :)
Then it becomes another issue. If the sensor keeps sending the invalid values for long time, what should zha display? Discard the update and show the previous value?

**Reply by mdeweerd** [reply]
I think that ZigPy should at least provide a means to let the application level identify the invalid values and distinguish them from unavailable values.
Then the application level has to evolve to use this new information.  

IMHO when the value is invalid, the application should not show the previous value: the current value is simply invalid.  In most cases it is as if it is not available, but the difference is that the ZED communicated something.

The application could trigger an event if the value is invalid for a long time.  I do not think that it is up to ZigPy.
